### PR TITLE
[5.1] Add Mandrill subaccounts

### PR DESCRIPTION
**Pull request by micbenner**
Am an open source virgin, go easy xP

Related to #5567

Really needed this so thought I'd give it a go.

Works fully without BC breaks

### CODE REVIEW

**Comment by micbenner** [comment]
Sorry, fixed CS

File: src/Illuminate/Mail/Transport/MandrillTransport.php [hunk 1]
```diff
@@ -22,16 +22,25 @@ class MandrillTransport extends Transport
     protected $key;
 
     /**
+     * The Mandrill subaccount.
+     *
+     * @var bool
+     */
+    protected $subaccount;
```
Comment (crynobone) [inline src/Illuminate/Mail/Transport/MandrillTransport.php]:
```php
protected $subaccount = false;
```

File: src/Illuminate/Mail/Transport/MandrillTransport.php [hunk 2]
```diff
@@ -22,16 +22,25 @@ class MandrillTransport extends Transport
...
      * @param  string  $key
+     * @param  bool    $subaccount
      * @return void
      */
-    public function __construct(ClientInterface $client, $key)
+    public function __construct(ClientInterface $client, $key, $subaccount = false)
     {
         $this->client = $client;
         $this->key = $key;
+        $this->subaccount = $subaccount;
```
Comment (crynobone) [inline src/Illuminate/Mail/Transport/MandrillTransport.php]: Don't think we should add boolean parameter here.

Why not

```php
public function enableSubAccount()
{
    $this->subaccount = true;
   
    return $this;
}

public function disableSubAccount()
{
    $this->subaccount = false;
   
    return $this;
}
```

**Comment by micbenner** [comment]
Noo, because the subaccount is either false or contains a string (to be set in the application's configuration file) defining what subaccount to use.

File: src/Illuminate/Mail/Transport/MandrillTransport.php [hunk 3]
```diff
@@ -41,6 +50,10 @@ public function send(Swift_Mime_Message $message, &$failedRecipients = null)
     {
         $this->beforeSendPerformed($message);
 
+        if ($this->subaccount !== false) {
+            $message->getHeaders()->addTextHeader('X-MC-Subaccount', $this->subaccount);
+        }
```
Comment (crynobone) [inline src/Illuminate/Mail/Transport/MandrillTransport.php]:
```php
if (! is_null($this->subaccount)) {
    $message->getHeaders()->addTextHeader('X-MC-Subaccount', $this->subaccount);
}
```

File: src/Illuminate/Mail/TransportManager.php [hunk 4]
```diff
@@ -113,7 +113,11 @@ protected function createMandrillDriver()
         $client = new HttpClient;
         $config = $this->app['config']->get('services.mandrill', []);
 
-        return new MandrillTransport($client, $config['secret']);
+        if (empty($config['subaccount'])) {
+            return new MandrillTransport($client, $config['secret']);
+        }
+
+        return new MandrillTransport($client, $config['secret'], $config['subaccount']);
```
Comment (crynobone) [inline src/Illuminate/Mail/TransportManager.php]:
```php
$mandrill = new MandrillTransport($client, $config['secret']);

if (! empty($config['subaccount'])) {
    $mandrill->setSubAccount($config['subaccount']);
}

return $mandrill;
```

**Comment by crynobone** [comment]
> Noo, because the subaccount is either false or contains a string (to be set in the application's configuration file) defining what subaccount to use.

I have updated my reply.

**Comment by micbenner** [comment]
Fair enough, will update to use null once I am back on my laptop.

Disagree with using a setter though, it is unnecessary and seems inconsistent with the other transport methods.

**Comment by crynobone** [comment]
> Disagree with using a setter though, it is unnecessary and seems inconsistent with the other transport methods.

Mailgun also has `setKey()` and `setDomain()` https://github.com/laravel/framework/blob/5.1/src/Illuminate/Mail/Transport/MailgunTransport.php#L136

And while you can say that this is non BC, someone might already extend `MandrillTransport` to add other feature (with Laravel, or used outside of Laravel), and could override the default constructor which will be breaking BC if this changes is merged.

**Comment by micbenner** [comment]
> And while you can say that this is non BC, someone might already extend ```MandrillTransport``` to add other feature

In which case they would be have overwritten the method in question to instantiate their child class. This would either use the parent constructor or their classes' own constructor. Neither of which would break backwards compatibility.

Your example of Mailgun uses the constructor to inject both key and domain values, which is the same method used as the other methods in ```TransportManager```

**Comment by crynobone** [comment]
> Your example of Mailgun uses the constructor to inject both key and domain values, which is the same method used as the other methods in TransportManager

Because it was introduce not in a patch release, so adding those in the constructor is acceptable since BC is not an issue, but you're asking to add additional parameter on a patch release.

Why not send this to `5.2` since BC wouldn't be an issue by then?

**Comment by micbenner** [comment]
Dude, it has a default value. It would not break anything.

**Comment by crynobone** [comment]
> Dude, it has a default value. It would not break anything.

What if someone already extends `MandrillTransport` (as I mentioned before), and yes they can.

```php
<?php 

use Illuminate\Mail\Transport\MandrillTransport;

class FooMandrillTransport extends MandrillTransport
{
    public function __construct(ClientInterface $client, $key, $someOtherOption, $somethingElse, $subAccount = null)
    {

         parent::__construct($client, $key);

         // set own properties.

    }
}
```

Not breaking?

**Comment by micbenner** [comment]
Tell me where your code would break.

**Comment by taylorotwell** [comment]
Sounds like we can't do this on 5.1 as it would be a breaking change.

**Comment by micbenner** [comment]
Would it be okay for me to submit this to the master branch then?

**Comment by taylorotwell** [comment]
Sure give it a shot.

**Pull request closed without merging**